### PR TITLE
[ACCEL] Add to ora2 return assessment flow

### PR DESCRIPTION
**Pull request by arsentur**
**Description:**
- Create a new button in the staff area for returning assessment flow.
- Create js handlers for this button
- Create xblock handler for processing returning assessment
- Create a model for storing returning assessment data
- Add new step and change existing workflow
- Provide returning step for students if their assessment was returned by staff

**YouTrack:**
https://youtrack.raccoongang.com/issue/ACCEL-197

### CODE REVIEW

File: openassessment/workflow/models.py [hunk 1]
```diff
@@ -65,7 +65,8 @@ class AssessmentWorkflow(TimeStampedModel, StatusModel):
                     # graded yet -- we're waiting for assessments of their
                     # submission by others.
         "done",  # Complete
-        "cancelled"  # User submission has been cancelled.
+        "cancelled",  # User submission has been cancelled.
+        "returned"  # User submission has been returned.
```
Comment (flying-pi) [inline openassessment/workflow/models.py]: Could you add a trailing comma here?

Comment (flying-pi) [inline openassessment/workflow/models.py]: Could you add trailing comma here?

File: openassessment/workflow/models.py [hunk 2]
```diff
@@ -311,6 +312,9 @@ def update_from_assessments(self, assessment_requirements, override_submitter_re
         if self.status == self.STATUS.cancelled:
             return
 
+        if self.status == self.STATUS.returned:
```
Comment (flying-pi) [inline openassessment/workflow/models.py]: Seems we can modify the previous condition to checking, that status in some list.

File: openassessment/workflow/models.py [hunk 3]
```diff
@@ -535,6 +539,33 @@ def cancel(self, assessment_requirements):
                 )
             )
 
+    def workflow_return(self, assessment_requirements):
```
Comment (flying-pi) [inline openassessment/workflow/models.py]: How we use `assessment_requirements` argument in the given function?

File: openassessment/workflow/models.py [hunk 4]
```diff
@@ -812,3 +879,65 @@ def get_latest_workflow_cancellation(cls, submission_uuid):
         """
         workflow_cancellations = cls.objects.filter(workflow__submission_uuid=submission_uuid).order_by("-created_at")
         return workflow_cancellations[0] if workflow_cancellations.exists() else None
+
+
+class AssessmentWorkflowReturning(models.Model):
+    """Model for tracking returning of assessment workflow.
```
Comment (flying-pi) [inline openassessment/workflow/models.py]: For me better, when doc strings start from new line ;)

File: openassessment/xblock/staff_area_mixin.py [hunk 5]
```diff
@@ -460,6 +460,33 @@ def cancel_submission(self, data, suffix=''):  # pylint: disable=W0613
 
         return self._cancel_workflow(submission_uuid, comments)
 
+    @XBlock.json_handler
+    @require_course_staff("STUDENT_INFO", with_json_handler=True)
+    def return_submission(self, data, suffix=''):  # pylint: disable=W0613
```
Comment (flying-pi) [inline openassessment/xblock/staff_area_mixin.py]: > pylint: 

Seems this project using pylint. Are you tried to use it?

Comment (flying-pi) [reply openassessment/xblock/staff_area_mixin.py]: Could you replace pylint exception code by it string representation?

File: openassessment/xblock/staff_area_mixin.py [hunk 6]
```diff
@@ -495,3 +522,39 @@ def _cancel_workflow(self, submission_uuid, comments, requesting_user_id=None):
...
             return {"success": False, 'msg': msg}
+
+    def _return_workflow(self, submission_uuid, comments, requesting_user_id=None):
+        """
+        Internal helper method to return a workflow using the workflow API.
+
+        If requesting_user is not provided, we will use the user to which this xblock is currently bound.
+        """
+        # Import is placed here to avoid model import at project startup.
+        from openassessment.workflow import api as workflow_api
```
Comment (flying-pi) [inline openassessment/xblock/staff_area_mixin.py]: Why we place import here, instead of the top of the file?

File: openassessment/xblock/staff_area_mixin.py [hunk 7]
```diff
@@ -495,3 +522,39 @@ def _cancel_workflow(self, submission_uuid, comments, requesting_user_id=None):
...
+        try:
+            assessment_requirements = self.workflow_requirements()
+            if requesting_user_id is None:
+                "The student_id is actually the bound user, which is the staff user in this context."
```
Comment (flying-pi) [inline openassessment/xblock/staff_area_mixin.py]: Is it comment?

File: openassessment/xblock/staff_area_mixin.py [hunk 8]
```diff
@@ -495,3 +522,39 @@ def _cancel_workflow(self, submission_uuid, comments, requesting_user_id=None):
             msg = ex.message
             logger.exception(msg)
             return {"success": False, 'msg': msg}
+
+    def _return_workflow(self, submission_uuid, comments, requesting_user_id=None):
```
Comment (flying-pi) [inline openassessment/xblock/staff_area_mixin.py]: this method is very similar to _cancel_workflow. Are you thinking about some kind of generalization?

File: openassessment/xblock/submission_mixin.py [hunk 9]
```diff
@@ -257,8 +259,22 @@ def create_submission(self, student_item_dict, student_sub_data, files_descripti
                 else:
                     break
 
+        workflow = self.get_workflow_info()
+
         submission = api.create_submission(student_item_dict, student_sub_dict)
-        self.create_workflow(submission["uuid"])
+
+        if workflow.get("status") == 'returned':
+            update_workflow = AssessmentWorkflow.objects.get(submission_uuid=workflow.get('submission_uuid'))
```
Comment (flying-pi) [inline openassessment/xblock/submission_mixin.py]: Seems `update` method from the Django ORM is more suitable here.

File: openassessment/xblock/workflow_mixin.py [hunk 10]
```diff
@@ -201,3 +201,26 @@ def get_workflow_cancellation_info(self, submission_uuid):
...
+        """
+        returning_info = workflow_api.get_assessment_workflow_returning(submission_uuid)
+        if not returning_info:
+            return None
+
+        # Add the username of the staff member who returned the submission
+        returning_info['returned_by'] = self.get_username(returning_info['returned_by_id'])
+
+        # Add the date that the workflow was returned (in preference to the serialized date string)
+        del returning_info['created_at']
```
Comment (flying-pi) [inline openassessment/xblock/workflow_mixin.py]: Why we drop `created_at`?

**Comment by flying-pi** [comment]
@arsentur What about some style checkers and running test? Do you investigated this area?

File: openassessment/workflow/models.py (hunk 1 above)
Comment (arsentur) [reply openassessment/workflow/models.py]: Done

File: openassessment/workflow/models.py (hunk 2 above)
Comment (arsentur) [reply openassessment/workflow/models.py]: Done

File: openassessment/xblock/staff_area_mixin.py (hunk 8 above)
Comment (arsentur) [reply openassessment/xblock/staff_area_mixin.py]: I think methods should be separated. because if in the future we would change return flow we can do it faster. It will be more flexible.

File: openassessment/workflow/models.py [hunk 11]
```diff
@@ -61,12 +61,12 @@ class AssessmentWorkflow(TimeStampedModel, StatusModel):
     STEPS = ASSESSMENT_API_DICT.keys()
 
     STATUSES = [
-        "waiting",  # User has done all necessary assessment but hasn't been
-                    # graded yet -- we're waiting for assessments of their
-                    # submission by others.
-        "done",  # Complete
+        "waiting",    # User has done all necessary assessment but hasn't been
```
Comment (idegtiarov) [inline openassessment/workflow/models.py]: Why these changes were done?

Comment (arsentur) [reply openassessment/workflow/models.py]: I added a comma after status

File: openassessment/xblock/staff_area_mixin.py (hunk 8 above)
Comment (flying-pi) [reply openassessment/xblock/staff_area_mixin.py]: Ok

File: openassessment/workflow/models.py (hunk 11 above)
Comment (idegtiarov) [reply openassessment/workflow/models.py]: the comma is great, why did you add two more spaces before the comment?

Comment (arsentur) [reply openassessment/workflow/models.py]: Done

**Comment by idegtiarov** [comment]
What about the tech debts, are you going to update the current branch with the upstream commits related to the hawthorn OpenEdx version?

**Comment by arsentur** [comment]
Yes, we will do it in the next PR
